cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/ProgramStateWriter.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.app.runtime;
+
+import co.cask.cdap.proto.BasicThrowable;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.id.ProgramRunId;
+
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * An interface that defines the behavior for how program states are persisted
+ */
+public interface ProgramStateWriter {
+
+  /**
+   * Updates the program run's status to be {@link ProgramRunStatus#STARTING} at the given start time
+   *
+   * @param programRunId the id of the program run
+   * @param twillRunId the run id of the twill application
+   * @param startTime the start time of the program in milliseconds when it has reached
+   *                  {@link ProgramRunStatus#STARTING}
+   */
+  void start(ProgramRunId programRunId, @Nullable String twillRunId, long startTime);
+
+  /**
+   * Updates the program run's status to be {@link ProgramRunStatus#RUNNING} at the given start time in seconds
+   *
+   * @param programRunId the id of the program run
+   * @param twillRunId the run id of the twill application
+   * @param startTime the start time of the program in milliseconds when it has reached
+   *                           {@link ProgramRunStatus#RUNNING}
+   */
+  void running(ProgramRunId programRunId, @Nullable String twillRunId, long startTime);
+
+  /**
+   * Updates the program run's status to be terminated at the given time with the given run status
+   *
+   * @param programRunId the id of the program run
+   * @param endTime the end time of the program in milliseconds when it has terminated
+   * @param runStatus the final run status of the program
+   * @param cause the reason for the program run's failure, if the program terminated with an error
+   */
+  void stop(ProgramRunId programRunId, long endTime,
```

Comment:
The abstraction can be done through a different means. The reason is because retry strategy for publishing to TMS might be different than retry strategy to write to Store (which probably the subscriber will retry the entire poll->persist logic). Besides, for the subscriber, it's better for it to use a different interface so that it can do it's job in more efficient way (e.g. batch write in one transaction).